File: additiongame/submissions/wrong_answer/lukas.cpp
//Solution by lukasP (Lukáš Poláček)
//Implementation of "A Combinatorial Problem on Abelian Groups" by Marshall Hall, Jr.
#include <iostream>
#include <vector>
#include <set>
#include <numeric>

#include <cstdio>
#include <cassert>
using namespace std;

#define rep(i,a,b) for(__typeof(b) i=a; i<(b); ++i)

typedef vector<int> vi;
int n;
vi a, b, c, pb, ipb;
void print(vi v, vi ipb)
{
    rep(ii,0,n)
    {
        int i = ipb[ii];
        cout << (v[i] == 0 ? n : v[i]) << (ii + 1 < n ? " " : "\n");
    }
}
void fix(int j, int nj)
{
    rep(k,0,n)
    {
        set<int> w = {c[j], c[nj]};
        int alt1 = (j + b[j]) % n;
        int alt2 = (j + b[nj]) % n;
        if (w.count(alt1) || w.count(alt2))
        {
            int res = w.count(alt1) ? alt1 : alt2;
            if (res == alt2)
            {
                swap(b[j], b[nj]);
                swap(pb[j], pb[nj]);
            }
            if (c[nj] == res)
                swap(c[j], c[nj]);
            return;
        }

        int sol = (n + c[j] - b[j]) % n;
        swap(b[j], b[sol]);
        swap(pb[j], pb[sol]);
        swap(c[j], c[sol]);
        swap(c[j], c[nj]);
    }
}
int main()
{
    scanf("%d", &n);
    a.resize(n); b.resize(n); c.resize(n);
    pb.resize(n); ipb.resize(n);

    rep(i,0,n)
    {
        a[i] = c[i] = pb[i] = i;
        scanf("%d", &b[i]);
    }

    if (accumulate(b.begin(), b.end(), 0) % n != 0)
    {
        cout << "impossible" << endl;
        return 0;
    }

    rep(i,0,n - 1)
        fix(i, i + 1);

    rep(j,0,n) ipb[pb[j]] = j;
    rep(j,0,n) a[j] = (n - a[j]) % n;

    print(a, ipb);
    print(c, ipb);
}

File: additiongame/submissions/accepted/lukas.cs
//Solution by lukasP (Lukáš Poláček)
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Diagnostics;

class Problem
{
    static int n;
    static List<int> a, b, c, pb;
    static String Print(List<int> a, List<int> p)
    {
        return String.Join(" ", Enumerable.Range(0, n).Select(x => a[p[x]]).Select(y => (y == 0 ? n : y)));
    }
    static void swap(List<int> x, int i, int j)
    {
        int tmp = x[i];
        x[i] = x[j];
        x[j] = tmp;
    }
    static void fix(int j, int nj)
    {
        while (true)
        {
            HashSet<int> w = new HashSet<int>(new int[]{c[j], c[nj]});
            int alt1 = (j + b[j]) % n;
            int alt2 = (j + b[nj]) % n;
            if (w.Contains(alt1) || w.Contains(alt2))
            {
                int res = w.Contains(alt1) ? alt1 : alt2;
                if (res == alt2)
                {
                    swap(b, j, nj);
                    swap(pb, j, nj);
                }
                if (c[nj] == res)
                    swap(c, j, nj);
                return;
            }

            int sol = (n + c[j] - b[j]) % n;
            swap(b, j, sol);
            swap(pb, j, sol);
            swap(c, j, sol);
            swap(c, j, nj);
        }
    }
    public static void Main ()
    {
        var bs = new BufferedStream(Console.OpenStandardInput());
        StreamReader reader = new StreamReader(bs);
        n = int.Parse(reader.ReadLine());

        a = Enumerable.Range(0, n).ToList();
        b = Enumerable.Repeat(0, n).ToList();
        c = Enumerable.Range(0, n).ToList();
        List <int> d = reader.ReadLine().Split().Select(x => int.Parse(x)).ToList();
        if (d.Sum() % n != 0)
        {
            Console.WriteLine("impossible");
            return;
        }

        pb = Enumerable.Range(0, n).ToList();
        List<int> ipb = Enumerable.Repeat(0, n).ToList();

        int s = 0;
        for (int i = 0; i < n - 1; i++)
        {
            for (int j = 0; j < n; j++)
                ipb[pb[j]] = j;
            s = (s + d[i]);
            b[ipb[i]] = d[i];
            b[ipb[i + 1]] = (n - s) % n;
            fix(ipb[i], ipb[i + 1]);
        }

        for (int i = 0; i < n; i++)
            Debug.Assert((a[i] + b[i]) % n == c[i]);
        for (int j = 0; j < n; j++)
            ipb[pb[j]] = j;

        Console.WriteLine(Print(a.Select(x => (n - x) % n).ToList(), ipb));
        Console.WriteLine(Print(c, ipb));
    }
}
